### PR TITLE
8300527: [Lilliput] Simplify and optimize loading of Klass*

### PR DESCRIPTION
**Pull request by rkennke**
With the new fast-locking, we can greatly simplify the loading of the Klass* from an object. We no longer have to call into the runtime to ensure that the header is changed under our feet (by installing the INFLATION marker). Instead, we can do a simple check for the object being monitor-locked, and safely load the displaced header, if that is the case.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8300527](https://bugs.openjdk.org/browse/JDK-8300527): [Lilliput] Simplify and optimize loading of Klass*

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - Committer) ⚠️ Review applies to [83c65c16](https://git.openjdk.org/lilliput/pull/66/files/83c65c16f3abaaa1995fa6cf44b103b751cc405d)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/lilliput pull/66/head:pull/66` \
`$ git checkout pull/66`

Update a local copy of the PR: \
`$ git checkout pull/66` \
`$ git pull https://git.openjdk.org/lilliput pull/66/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 66`

View PR using the GUI difftool: \
`$ git pr show -t 66`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/lilliput/pull/66.diff">https://git.openjdk.org/lilliput/pull/66.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/c1/c1_LIR.cpp [hunk 1]
```diff
@@ -888,7 +888,7 @@ void LIR_OpVisitState::visit(LIR_Op* op) {
 
       do_input(opLoadKlass->_obj);
       do_output(opLoadKlass->_result);
-      do_stub(opLoadKlass->_stub);
+      do_stub(opLoadKlass->stub());
```
Comment (shipilev) [inline src/hotspot/share/c1/c1_LIR.cpp:891]: Minor: why not leave `_stub`?

File: src/hotspot/share/runtime/synchronizer.cpp [hunk 2]
```diff
@@ -795,7 +795,7 @@ struct SharedGlobals {
 
 static SharedGlobals GVars;
 
-markWord ObjectSynchronizer::read_stable_mark(oop obj) {
+static markWord read_stable_mark(oop obj) {
```
Comment (shipilev) [inline src/hotspot/share/runtime/synchronizer.cpp:798]: Do you still need this method, or? If you still need it, let's not put it outside the `ObjectSynchronizer` "namespace".

**Review by shipilev — APPROVED** [review]
Minor nits.

File: src/hotspot/share/c1/c1_LIR.cpp (hunk 1 above)
Comment (rkennke) [reply src/hotspot/share/c1/c1_LIR.cpp]: Right. I changed that back.

File: src/hotspot/share/runtime/synchronizer.cpp (hunk 2 above)
Comment (rkennke) [reply src/hotspot/share/runtime/synchronizer.cpp:798]: Originally, read_stable_mark() is static in upstream. A while back I put it into the ObjectSynchronizer namespace to reuse it in Shenandoah, but that is no longer needed, so I reverted the scope back to compilation-unit-private. Yes, we will still need it, when we make Lilliput switchable, but I'd keep it compilation-unit-private for the reasons mentioned.

Comment (shipilev) [reply src/hotspot/share/runtime/synchronizer.cpp:798]: All right then!

**Pull request closed without merging**